### PR TITLE
blocks: use auto-generated .c files instead of 'LD -r -b binary'

### PR DESCRIPTION
**Pull request by xiphon**
Rebased #4472  + Fixes OSX compilation

### CODE REVIEW

**Comment by xiphon** [comment]
Rebased, ready to merge

**Comment by xiphon** [comment]
@moneromooo-monero 
which one?

**Comment by moneromooo-monero** [comment]
Weird, I'm pretty sure I typed a review comment. It's about making the std::function parameters const refs, as they can get largeish.

**Comment by moneromooo-monero** [comment]
There's at least a couple which are still not, functions in core and Blockchain.

**Comment by xiphon** [comment]
Kinda hard to guess what lines we are talking about. Feel free to comment on these lines.

File: src/cryptonote_core/blockchain.cpp [hunk 1]
```diff
@@ -341,7 +338,7 @@ uint64_t Blockchain::get_current_blockchain_height() const
 //------------------------------------------------------------------
 //FIXME: possibly move this into the constructor, to avoid accidentally
 //       dereferencing a null BlockchainDB pointer
-bool Blockchain::init(BlockchainDB* db, const network_type nettype, bool offline, const cryptonote::test_options *test_options, difficulty_type fixed_difficulty)
+bool Blockchain::init(BlockchainDB* db, const network_type nettype, bool offline, const cryptonote::test_options *test_options, difficulty_type fixed_difficulty, const GetCheckpointsCallback get_checkpoints/* = nullptr*/)
```
Comment (moneromooo-monero) [inline src/cryptonote_core/blockchain.cpp]: here

File: src/cryptonote_core/blockchain.h [hunk 2]
```diff
@@ -117,10 +128,11 @@ namespace cryptonote
      * @param offline true if running offline, else false
      * @param test_options test parameters
      * @param fixed_difficulty fixed difficulty for testing purposes; 0 means disabled
+     * @param get_checkpoints if set, will be called to get checkpoints data
      *
      * @return true on success, false if any initialization steps fail
      */
-    bool init(BlockchainDB* db, const network_type nettype = MAINNET, bool offline = false, const cryptonote::test_options *test_options = NULL, difficulty_type fixed_difficulty = 0);
+    bool init(BlockchainDB* db, const network_type nettype = MAINNET, bool offline = false, const cryptonote::test_options *test_options = NULL, difficulty_type fixed_difficulty = 0, const GetCheckpointsCallback get_checkpoints = nullptr);
```
Comment (moneromooo-monero) [inline src/cryptonote_core/blockchain.h]: here

File: src/cryptonote_core/blockchain.h [hunk 3]
```diff
@@ -1369,8 +1381,10 @@ namespace cryptonote
      * A (possibly empty) set of block hashes can be compiled into the
      * monero daemon binary.  This function loads those hashes into
      * a useful state.
+     * 
+     * @param get_checkpoints if set, will be called to get checkpoints data
      */
-    void load_compiled_in_block_hashes();
+    void load_compiled_in_block_hashes(const GetCheckpointsCallback get_checkpoints);
```
Comment (moneromooo-monero) [inline src/cryptonote_core/blockchain.h]: here

File: src/cryptonote_core/cryptonote_core.cpp [hunk 4]
```diff
@@ -389,7 +389,7 @@ namespace cryptonote
     return m_blockchain_storage.get_alternative_blocks_count();
   }
   //-----------------------------------------------------------------------------------------------
-  bool core::init(const boost::program_options::variables_map& vm, const char *config_subdir, const cryptonote::test_options *test_options)
+  bool core::init(const boost::program_options::variables_map& vm, const char *config_subdir, const cryptonote::test_options *test_options, const GetCheckpointsCallback get_checkpoints/* = nullptr */)
```
Comment (moneromooo-monero) [inline src/cryptonote_core/cryptonote_core.cpp]: here

File: src/cryptonote_core/cryptonote_core.h [hunk 5]
```diff
@@ -244,10 +244,11 @@ namespace cryptonote
       * @param vm command line parameters
       * @param config_subdir subdirectory for config storage
       * @param test_options configuration options for testing
+      * @param get_checkpoints if set, will be called to get checkpoints data, must return checkpoints data pointer and size or nullptr if there ain't any checkpoints for specific network type
       *
       * @return false if one of the init steps fails, otherwise true
       */
-     bool init(const boost::program_options::variables_map& vm, const char *config_subdir = NULL, const test_options *test_options = NULL);
+     bool init(const boost::program_options::variables_map& vm, const char *config_subdir = NULL, const test_options *test_options = NULL, const GetCheckpointsCallback get_checkpoints = nullptr);
```
Comment (moneromooo-monero) [inline src/cryptonote_core/cryptonote_core.h]: here

**Comment by xiphon** [comment]
Thanks, done

**Pull request merged**